smdebug/tensorflow/keras.py

```diff
@@ -263,26 +271,34 @@ def _is_output_layer(self, mode, layer_outputs):
                 model_outputs.extend(per_replica_model.outputs)
         else:
             model_outputs.extend(model.outputs)
+        # In TF 2.X, calling `layer_outputs[0] in model_outputs gives the error:
+        # *** tensorflow.python.framework.errors_impl.OperatorNotAllowedInGraphError: using a
+        # `tf.Tensor` as a Python `bool` is not allowed in Graph execution. Use Eager execution or
+        # decorate this function with @tf.function.
+        # Calling `layer_outputs[0] == model_outputs[0]` gives <tf.Tensor 'Equal_1:0'>
         return any([i in model_outputs for i in layer_outputs])
 
     def _prepare_layers(self, mode):
         # adds any layer tensor (input, output and weight) to appropriate collection
         for layer in self.model.layers:
-            layer_inputs = get_keras_layer_inputs(layer)
-            is_input_layer = self._is_input_layer(mode, layer_inputs)
-            for inp in layer_inputs:
-                self._check_and_add_layer_tensor(
-                    mode, layer, "input", inp, is_input_to_model=is_input_layer
-                )
+            # Input and output tensors are difficult to get in TF 2.X
+            if version.parse(tf.__version__) < version.parse("2.0.0"):
+                layer_inputs = get_keras_layer_inputs(layer)
+                is_input_layer = self._is_input_layer(mode, layer_inputs)
+                for inp in layer_inputs:
+                    self._check_and_add_layer_tensor(
+                        mode, layer, "input", inp, is_input_to_model=is_input_layer
+                    )
 
-            layer_outputs = get_keras_layer_outputs(layer)
+                layer_outputs = get_keras_layer_outputs(layer)
 
-            is_output_layer = self._is_output_layer(mode, layer_outputs)
-            for outp in layer_outputs:
-                self._check_and_add_layer_tensor(
-                    mode, layer, "output", outp, is_output_of_model=is_output_layer
-                )
+                is_output_layer = self._is_output_layer(mode, layer_outputs)
+                for outp in layer_outputs:
+                    self._check_and_add_layer_tensor(
+                        mode, layer, "output", outp, is_output_of_model=is_output_layer
+                    )
 
+            # Weights can be retrieved in both
             weights = layer.weights
             for w in weights:
                 self._check_and_add_layer_tensor(mode, layer, "weight", w)
```

Comment:
Doesn't this duplicate some of the work done in _save_tensors_post_step for the weights and biases?